ocaml/typing/env.ml

```diff
@@ -3145,40 +3211,47 @@ and lookup_all_args ~errors ~use ~loc lid0 env =
     | Lident _ | Ldot _ as f_lid ->
         (f_lid, args)
     | Lapply (f_lid, arg_lid) ->
-        let arg_path, arg_md = lookup_module ~errors ~use ~loc arg_lid env in
-        loop_lid_arg ((f_lid,arg_path,arg_md.md_type)::args) f_lid
+        let arg_path, arg_md, arg_vmode =
+          lookup_module ~errors ~use ~loc arg_lid env
+        in
+        loop_lid_arg ((f_lid,arg_path,arg_md.md_type,arg_vmode)::args) f_lid
   in
   loop_lid_arg [] lid0
 
 and lookup_apply ~errors ~use ~loc lid0 env =
   let f0_lid, args0 = lookup_all_args ~errors ~use ~loc lid0 env in
-  let args_for_errors = List.map (fun (_,p,mty) -> (p,mty)) args0 in
-  let f0_path, f0_comp =
+  let args_for_errors = List.map (fun (_,p,mty,vmode) -> (p,mty,vmode.mode)) args0 in
+  let f0_path, f0_locks, f0_comp =
     lookup_module_components ~errors ~use ~loc f0_lid env
   in
-  let check_one_apply ~errors ~loc ~f_lid ~f_comp ~arg_path ~arg_mty env =
+  if use then begin
+    walk_locks ~errors ~loc ~env ~item:Module ~lid:f0_lid mda_mode None f0_locks
+    |> ignore
```

Comment:
I think that `lookup_apply` is only ever used when looking up types and module types. At the moment I think you will error for this:
```ocaml
module M = struct end
module F (M : sig end) = struct type t end
let use_portable (f @ portable) = f ()
let _ = use_portable (fun () -> let module _ = struct type t = F(M).t end in ())
```
I think you can just not walk the locks here.

Reply:
Yes you are right. I looked at `parser.mly` and `Lapply` can only be generated for "compile-time" module application. For runtime module application (where a pointer to the functor is needed), `Pmod_apply` is used instead.

I was under the wrong impression that we can reuse the `~use` argument is to distinguish between the two usages. This is also why in a previous discussion I insisted that `Env.lookup_* ~use:true` is a "real" usage and should just walk the locks and enforcing legacy. This is now obviously wrong for modules.

Reply:
Well in the example, the look-up of `M` must also not cause walking locks. The proper way is to add `~lock` to `lookup_module` and siblings. I believe module is the only thing that needs it - for other categories, `~use:true` means needing walking locks.

Reply:
Ok there are some other cases where a reference to a module should not trigger locks. This is fixed and tested in the commit "some ref to module should not trigger locks".